### PR TITLE
Labels generation

### PR DESCRIPTION
**Pull request by bartoszLesniewski**
Added generating labels with QR codes for items and collections. Some useful information about mechanisms used can be found on [our confluence page](https://engineering-project.atlassian.net/wiki/spaces/SD/pages/34308097/Labels+generation)

### CODE REVIEW

File: src/Form/Type/Entity/LabelType.php [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        ;
+    }
+
+    public function configureOptions(OptionsResolver $resolver): void
+    {
+        $resolver->setDefaults([
+            'data_class' => Label::class,
+        ]);
+    }
+}
```
Comment (punfil) [inline src/Form/Type/Entity/LabelType.php]: Please add new line at the end of file ;)

File: src/Enum/LabelSizeEnum.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+namespace App\Enum;
+
+class LabelSizeEnum
+{
+    public const SIZE_A4 = 'A4';
+
+    public const SIZE_A5 = 'A5';
+
+    public const SIZE_A6 = 'A6';
```
Comment (punfil) [inline src/Enum/LabelSizeEnum.php:13]: I believe for labels very common size is A6P (at least for shipping ones), so I propose to add support for that size.

Please consider.

File: src/Service/LabelsGenerator.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+        $dompdf->loadHtml($htmlContent);
+        $dompdf->render();
+
+        $filename = "item_label_{$labelSize}.pdf";
+
+        return array('content' => $dompdf->output(),
+                     'filename' => $filename);
+    }
+}
```
Comment (punfil) [inline src/Service/LabelsGenerator.php]: Here as well ;)

File: templates/App/Label/label.html.twig [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+                    <p>
+                       <b>{{ datum.label }}:</b> {{ datum.value }}
+                    </p>
+                {% endfor %}
+            </div>
+            <div class="qrCode">
+                <img src="{{ qrCode }}">
+            </div>
+        </div>
+
```
Comment (punfil) [inline templates/App/Label/label.html.twig]: no need for this extra line?

Please consider

File: src/Form/Type/Entity/LabelType.php (hunk 1 above)
Comment (bartoszLesniewski) [reply src/Form/Type/Entity/LabelType.php]: Done

File: src/Enum/LabelSizeEnum.php (hunk 2 above)
Comment (bartoszLesniewski) [reply src/Enum/LabelSizeEnum.php:13]: I checked and A6P has the same dimensions as A6 size

File: src/Service/LabelsGenerator.php (hunk 3 above)
Comment (bartoszLesniewski) [reply src/Service/LabelsGenerator.php]: Done

File: templates/App/Label/label.html.twig (hunk 4 above)
Comment (bartoszLesniewski) [reply templates/App/Label/label.html.twig]: Why not? The same rules as in php files

**Review by punfil — COMMENTED** [review]
Verified locally with printing on A6 sized paper:
- Please set margins to 0, because printing on A6 doesn't work properly otherwise, because the label is shifted.
- Please include item name in the center of the top side (I suppose) to have item's name on the label. I propose to make it bold and larger then the rest of the text. I am not sure about collection name (quantity it not important I think, at least in the exhibition).

**Review by punfil — COMMENTED** [review]
The PR looks great, thanks for delivering the feature.
However, I still some comments to consider.
1. Icons order. For collection, generate-label is the first icon, but I think should be 4th as in item - after edit, which makes sense.
2. I would propose to add mode, where I can generate label for multiple items at once.
So when I am in collection view, I want to first choose if the label would be for collection or labels for each item selected.
3. Translation of */generate-label.
4. I've tested multiple fonts and sizes and managed to bug the generator:
a. When using long text, the label splits into multiple pages. I propose to limit text if it exceeds given threshold or unsupport long text fields or any better solution you can come up with.
b. When using multiple fields in specific scenarios, the QR code overflows the text on the label.
c. Option to enlarge the QR code along with font size (but this is more like feature req and I don't know if that makes sense?)
d. I think user should be able to choose which parts of item description he wants to have on the label.
e. in horizontal mode, item name is not centered.
f. I propose check for null values. Let's say item has "Informacje dodatkowe", which is empty, I don't want to have that on the label. However, I like that it doesn't print private fields.
 
Please consider.

**Comment by Lenetis** [comment]
I checked how the real labels on the exhibition look, and I have a few suggestions:

1. Most of the physical labels have their text centered (both horizontally and vertically). It would be nice to generate centered labels as well, or maybe have an option to toggle that.
2. The QR code doesn't need to be so large in most cases. Maybe add an option to set a custom size?
3. I would make the QR code optional - in case of small labels there isn't enough space for it, and not every item will have additional information on the website.
4. In some cases, it would be nice to be able to specify which fields to include on the label (e.g. only print short specifications, with additional information available on the website). Probably difficult to implement fully, but maybe excluding long text fields would be enough?
5. The label generates with a black border - why?
6. Physical labels have their name printed as a first field, not a header (i.e. "Nazwa: nazwa przedmiotu"). I like your solution better, but just wanted to point it out in case you missed it.

+1 to all the points mentioned in the previous comment

File: src/Enum/OrientationEnum.php [hunk 5]
```diff
@@ -17,6 +17,10 @@ class OrientationEnum
 
     public static function getOrientations(): array
     {
-        return self::ORIENTATIONS;
+        return [
+            self::ORIENTATION_VERTICAL => 'global.generate_label.orientation.vertical',
```
Comment (punfil) [inline src/Enum/OrientationEnum.php:21]: This should be corrected in `self::ORIENTATIONS` instead? Or remove `self::ORIENTATIONS`, because it's used nowhere?

Please consider

**Review by punfil — APPROVED** [review]

File: src/Enum/OrientationEnum.php [hunk 6]
```diff
@@ -17,6 +17,10 @@ class OrientationEnum
 
     public static function getOrientations(): array
     {
-        return self::ORIENTATIONS;
+        return [
```
Comment (punfil) [inline src/Enum/OrientationEnum.php]: self::ORIENTATIONS is no longer required?

**Comment by bartoszLesniewski** [comment]
> The PR looks great, thanks for delivering the feature. However, I still some comments to consider.
> 
> 1. Icons order. For collection, generate-label is the first icon, but I think should be 4th as in item - after edit, which makes sense.
> 2. I would propose to add mode, where I can generate label for multiple items at once.
>    So when I am in collection view, I want to first choose if the label would be for collection or labels for each item selected.
> 3. Translation of */generate-label.
> 4. I've tested multiple fonts and sizes and managed to bug the generator:
>    a. When using long text, the label splits into multiple pages. I propose to limit text if it exceeds given threshold or unsupport long text fields or any better solution you can come up with.
>    b. When using multiple fields in specific scenarios, the QR code overflows the text on the label.
>    c. Option to enlarge the QR code along with font size (but this is more like feature req and I don't know if that makes sense?)
>    d. I think user should be able to choose which parts of item description he wants to have on the label.
>    e. in horizontal mode, item name is not centered.
>    f. I propose check for null values. Let's say item has "Informacje dodatkowe", which is empty, I don't want to have that on the label. However, I like that it doesn't print private fields.
> 
> Please consider.

1. Done
2. Done
3. Done
4. Done
a. Fixed - now text doesn't go to the next page, but it is trimmed when exceeds its area
b. This is intentional behaviour, as label is divided into 2 parts - "content" area (by dafault 75%) where text is located and "QR code" area (by default 25%) where there is an image with QR code. If text exceeds its area, it is simply hidden and it looks like it is covered by QR code (that is connected with point 4a). Alternatively, we can use JavaScript to autoscale text, but I'm not sure if it would work properly with DOMPdf (and also I guess it would take quite a lot of work)
c. Done
d. Done
e. It is centered, but in the section which contains text. I would prefer to leave it as it is to make the template more flexible and easier to manage the arrangement of elements.
f. Done

File: src/Enum/OrientationEnum.php (hunk 6 above)
Comment (bartoszLesniewski) [reply src/Enum/OrientationEnum.php]: Moved translations to this array

**Comment by bartoszLesniewski** [comment]
> I checked how the real labels on the exhibition look, and I have a few suggestions:
> 
> 1. Most of the physical labels have their text centered (both horizontally and vertically). It would be nice to generate centered labels as well, or maybe have an option to toggle that.
> 2. The QR code doesn't need to be so large in most cases. Maybe add an option to set a custom size?
> 3. I would make the QR code optional - in case of small labels there isn't enough space for it, and not every item will have additional information on the website.
> 4. In some cases, it would be nice to be able to specify which fields to include on the label (e.g. only print short specifications, with additional information available on the website). Probably difficult to implement fully, but maybe excluding long text fields would be enough?
> 5. The label generates with a black border - why?
> 6. Physical labels have their name printed as a first field, not a header (i.e. "Nazwa: nazwa przedmiotu"). I like your solution better, but just wanted to point it out in case you missed it.
> 
> +1 to all the points mentioned in the previous comment

1. I added an option to select text alignment (left and center)
2. Now we set the size of QR Code as % of the whole label
3. If you set QR code size to 0%, it won't appear on a label
4. Done
5. Removed. It was used only for tests, just to control position of elements
6. Yes, I know... but header looks better in my opinion :D

File: templates/App/Collection/generate_label.html.twig [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+{% endblock %}
+
+{% block content %}
+    <div class="title-block">
+        <h1>{{ 'title.generate_label_collection'|trans({'collection': collection.title}) }}</h1>
+    </div>
+
+    <div class="content-block">
+        {{ form_start(form) }}
+        <h3 class="header col s12">Label settings</h3>
```
Comment (punfil) [inline templates/App/Collection/generate_label.html.twig]: I think this should be translated. Please follow for fields and items as well.

File: templates/App/Collection/generate_label.html.twig [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                {% else %}
+                    <div class="col s12 l5 offset-l1 xl4 offset-xl2">
+                {% endif %}
+                    {{ form_row(item, {'attr': {'class': 'itemCheck'}}) }}
+                </div>
+            {% endfor %}
+        </div>
+        {% endif %}
+
+        <h3 class="header col s12">Selection of fields</h3>
```
Comment (punfil) [inline templates/App/Collection/generate_label.html.twig]: I propose to display this only if collection has fields to avoid:
obraz

File: templates/App/Collection/generate_label.html.twig [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+        <div class="row">
+            <div class="col s12 l5 offset-l1 xl4 offset-xl2">
+                <strong>{{ form_row(form.selectAllFields, {'label' : 'label.labels_select_deselect_all', 'attr': {'class': 'selectAllFields'}}) }}</strong>
```
Comment (punfil) [inline templates/App/Collection/generate_label.html.twig]: It seems to not work.
When I unclicked all and clicked this one, the label has no fields.

When I click unlick all, other fields are still clicked. When I unclick any field, this shouldn't be checked IMO.

Please consider

**Comment by Lenetis** [comment]
When generating labels for items in a collection that has only one item, I got the following error:

image

When generating labels for only the second item of collection that has two items, I got the following error:

image

(Other cases that I tested worked correctly)

File: templates/App/Collection/generate_label.html.twig (hunk 7 above)
Comment (bartoszLesniewski) [reply templates/App/Collection/generate_label.html.twig]: Done

File: templates/App/Collection/generate_label.html.twig (hunk 8 above)
Comment (bartoszLesniewski) [reply templates/App/Collection/generate_label.html.twig]: Good point. Done

File: templates/App/Collection/generate_label.html.twig (hunk 9 above)
Comment (bartoszLesniewski) [reply templates/App/Collection/generate_label.html.twig]: It should work correctly now. Please re-review

**Comment by bartoszLesniewski** [comment]
> When generating labels for items in a collection that has only one item, I got the following error:
> 
> image
> 
> When generating labels for only the second item of collection that has two items, I got the following error:
> 
> image
> 
> (Other cases that I tested worked correctly)

Thanks for noticing these cases. Null values weren't handled correctly and there was also an issue with array keys when passing them from checkboxes in form to the labels generator. I fixed both. Please re-review

**Review by punfil — COMMENTED** [review]
I am not still convinced select all button works OK.
Discussed offline.

**Review by punfil — APPROVED** [review]
LGTM, thanks for the extensive work

**Review by Lenetis — APPROVED** [review]

**Review by punfil — APPROVED** [review]
Having tested the most recent commit, I have just one proposal. 
I've tested it in a scenario where there is one item, but has a very long content of fields.
Result:
obraz

Trimming seems OK to me, but I propose to end the grid.

**Comment by bartoszLesniewski** [comment]
> Having tested the most recent commit, I have just one proposal. I've tested it in a scenario where there is one item, but has a very long content of fields. Result: obraz
> 
> Trimming seems OK to me, but I propose to end the grid.

Unfortunately, it's not possible to end the grid when overflow occured (this is how hidden attribute works). Instead, we can move the overflowed part to the next page, but I'm not convinced that's a good idea. Btw. normally words are wrapped, you put one very long word in your example, that's why there is an overflow (idk if such a long word exists in any language xD)

File: docker-compose.dist.yml [hunk 10]
```diff
@@ -13,37 +13,36 @@ services:
         environment:
             - APP_ENV=dev
             - APP_DEBUG=1
-            - DB_DRIVER=pdo_pgsql
```
Comment (punfil) [inline docker-compose.dist.yml]: Please remove first commit as it shouldn't be part of this PR.
The rest LGTM.

**Review by punfil — APPROVED** [review]

File: docker-compose.dist.yml (hunk 10 above)
Comment (bartoszLesniewski) [reply docker-compose.dist.yml]: Done

**Review by punfil — APPROVED** [review]
